packages/client/tests/functional/methods/upsert/native-atomic/tests.ts

```diff
@@ -0,0 +1,241 @@
+import { faker } from '@faker-js/faker'
+// @ts-ignore
+import type { PrismaClient } from '@prisma/client'
+
+import { NewPrismaClient } from '../../../_utils/types'
+import testMatrix from './_matrix'
+
+declare let newPrismaClient: NewPrismaClient<typeof PrismaClient>
+
+// When awaited will return the latest UPDATE statement
+function createSQLPromise(client: PrismaClient) {
+  return ((): Promise<string> =>
+    new Promise((resolve) => {
+      // @ts-expect-error
+      client.$on('query', (data) => {
+        if ('query' in data && data.query.includes('UPDATE')) {
+          resolve(data.query)
+        }
+      })
+    }))()
+}
+
+testMatrix.setupTestSuite(
+  () => {
+    let client: PrismaClient
+
+    beforeAll(() => {
+      client = newPrismaClient({
+        log: [
+          {
+            emit: 'event',
+            level: 'query',
+          },
+        ],
+      })
+    })
+
+    test('should only use ON CONFLICT when update arguments do not have any nested queries', async () => {
+      const name = faker.name.firstName()
+      const title = faker.name.jobTitle()
+
+      await client.user.create({
+        data: {
+          name,
+          posts: {
+            create: {
+              title,
+            },
+          },
+        },
+      })
+
+      let sqlPromise: Promise<string> = createSQLPromise(client)
+      let doesSQLContainOnConflict = false
+
+      // This will 'not' use ON CONFLICT
+      await client.user.upsert({
+        where: {
+          name,
+        },
+        create: {
+          name,
+        },
+        update: {
+          name,
+          // Because this is a nested mutation
```

Comment:
Is this the only possible nested mutation? Should we maybe test "all" of them?

Reply:
I've added a few more different operations.